nulib/meadow
# Adapt Livebook's runtime configuration for Meadow

# Summary 

The `Hush` module made our configuration super complicated, and resulted in near constant compilation and runtime warnings about redefining modules. Livebook uses a [pretty good pattern](https://github.com/livebook-dev/livebook/blob/main/lib/livebook.ex#L84) for runtime configuration that I've tried to replicate here. It was tricky managing different environments that all have some level of dynamic/runtime config, or whose configuration might be overwritten by the top-level runtime config, but I think I came up with a reasonable middle ground. It's still a departure from standard Elixir practice, but it's better than it was.

# Specific Changes in this PR
- Remove `hush` and `hush_aws_secretsmanager` dependencies
- Create `Meadow.Runtime` module with convenience functions for retrieving AWS and environment secrets
- Create modules for the same kind of hierarchical runtime config that `config/*.exs` does for compile-time config:
  - `Meadow.Runtime.Config`
  - `Meadow.Runtime.{Dev, Test, Prod}`
  - `Meadow.Runtime.Release`
  - `Meadow.Runtime.Pipeline`
- Minimal updates to `lib` and `test` code to account for the new runtime config (mostly the disappearance of `Hush`)
- Update to `ueberauth_nusso` v2.0.0 to support a split between compile-time Ueberauth config and runtime Ueberauth strategy config

# Version bump required by the PR

See [Semantic Versioning 2.0.0](https://semver.org/) for help discerning which is required.


# Steps to Test

Just run it and make sure everything's behaving, and look over the new config modules to make sure we all understand where any new/changed config would go.

# :rocket: Deployment Notes

**Note** - __if you check any of these boxes go to the [(always open) `main` <- `staging` PR](https://github.com/nulib/meadow/pulls) and add detailed notes and instructions to help out others who may end up deploying your changes to production__

- Backward compatible API changes
- Backwards-incompatible API changes


# Tested/Verified
